ioSPI/fourier.py

```diff
@@ -0,0 +1,212 @@
+    only_real : bool
+        If True, return only real part of transform.
+
+    Returns
+    -------
+    arr3d_f : numpy.ndarray, shape (N,N,N)
+        Output array.
+    """
+    if neg_pos_3d is None:
+        neg_pos_3d = make_neg_pos_3d(arr3d)
+    if arr3d.shape[0] % 4 != 0:
+        neg_pos_3d *= -1
+
+    if mode == "f":
+        arr3d_f = numpy_fft.fftn(neg_pos_3d * arr3d)
+        arr3d_f /= np.sqrt(np.prod(arr3d_f.shape))
+    elif mode == "i":
+        arr3d_f = numpy_fft.ifftn(neg_pos_3d * arr3d)
+        arr3d_f *= np.sqrt(np.prod(arr3d_f.shape))
+
+    if only_real:
+        arr3d_f = arr3d_f.real
+
+    arr3d_f *= neg_pos_3d
+    return arr3d_f
+
+
+def do_fft(arr, d=3, only_real=False, **kwargs):
+    """FFT wrapper.
+
+    Wraps fft2d
+
+    Parameters
+    ----------
+    arr : numpy.ndarray, shape (N,N) or (N,N,N)
+        Input array.
+    d : int, 2 or 3
+        Dimension.
+    only_real : bool
+        If True, return only real part of transform.
+
+    Returns
+    -------
+    arr_f : numpy.ndarray, shape (N,N) or (N,N,N)
+        Output array.
+    """
+    if d == 2:
+        arr_f = fft2d(arr, mode="f", only_real=only_real, **kwargs)
+    elif d == 3:
+        arr_f = fft3d(arr, mode="f", only_real=only_real, **kwargs)
+    return arr_f
+
+
+def do_ifft(arr_f, d=3, only_real=True, **kwargs):
+    """iFFT wrapper.
+
+    Wraps fft2d
+
+    Parameters
+    ----------
+    arr_f : numpy.ndarray, shape (N,N) or (N,N,N)
+        Input array.
+    d : int, 2 or 3
+        Dimension.
+    only_real : bool
+        If True, return only real part of transform.
+
+    Returns
+    -------
+    arr : numpy.ndarray, shape (N,N) or (N,N,N)
+        Output array.
+    """
+    if d == 2:
+        arr = fft2d(arr_f, mode="i", only_real=only_real, **kwargs)
+    elif d == 3:
+        arr = fft3d(arr_f, mode="i", only_real=only_real, **kwargs)
+    return arr
+
+
+def fft2d(
+    arr2d,
+    mode,
+    numpy_fft=pyfftw.interfaces.numpy_fft,
+    only_real=False,
+    batch=False
+):
+    """2D FFT
+
+    Fast fourier transform
+    TODO: look into pyfftw.interfaces.numpy_fft.irfftn .
+
+    Parameters
+    ----------
+    arr2d : numpy.ndarray, shape (N,N) or (num_exposures,N,N)
+        Input array.
+    mode : str
+        Forward or reverse
+    numpy_fft : func
+        Function backend for performing the fft.
+    only_real : bool
+        If True, return only real part of transform.
+    batch : bool
+        Use for batch of exposures.
+
+    Returns
+    -------
+    arr2d_f : numpy.ndarray, shape (N,N) or (num_exposures,N,N)
+        Output array.
+    """
+    assert (arr2d.ndim == 2 and not batch) or (batch and arr2d.ndim == 3)
+    n1, n2 = arr2d.shape[-2:]
+    assert n1 == n2
+    # we apply an alterating +1/-1 multiplicative
+    # before we go to/from Fourier space.
+    # Later we apply this again to the transform.
+    arr2d = neg_pos_2d(arr2d.reshape(-1, n1, n1).copy())
```

Comment:
Maybe homogenize the way it is done in 2D and in 3D?

Reply:
A number of things are done differently, like the normalization, and if there is batch or not. The core functions fft2d and fft3d are different, but in the end there's a higher level wrapper `do_fft` and `do_ifft` where you just give an array and the dimension.